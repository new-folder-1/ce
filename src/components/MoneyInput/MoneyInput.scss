.MoneyInput {
    height: 32px;
    font-size: 20px;
    background-color: transparent;
    color: #fff;
    text-align: right;
    border: none;
    outline: none;
    width: 100%;

    &::placeholder {
        color: #fff;
        opacity: 0.3;
    }

    &[disabled] {
        opacity: 0.6;
    }
}
