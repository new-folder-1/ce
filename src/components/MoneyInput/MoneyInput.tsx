import * as React from 'react';

import { isValidNumber, clearLeadingZeros } from '../../utils';

import './MoneyInput.scss';

export interface MoneyInputProps {
    prefix: '+' | '-';
    amount?: number;
    autofocus?: boolean;
    onChange: (value: number) => void;
}

const addPrefix = (string: string, prefix: string) => `${prefix}${string}`;

export const MoneyInput =({
    prefix, amount, autofocus, onChange
} : MoneyInputProps) => {
    const [rawValue, setRawValue] = React.useState(amount ? amount.toString() : '');
    const [value, setValue] = React.useState(amount);

    const _prefix = prefix + ' ';

    React.useEffect(() => {
        if (amount === value) return;

        setValue(amount);
        const amountString = amount ? amount.toString() : '';
        setRawValue(amount ? addPrefix(amountString, _prefix) : amountString);
    }, [amount, value]);

    const onInput = (e: React.ChangeEvent<HTMLInputElement>) => {
        const inputVal = e.currentTarget.value;        
        let clearInputVal = inputVal.startsWith(_prefix) ? inputVal.substring(_prefix.length) : inputVal;

        if (!isValidNumber(clearInputVal)) {
            return;
        }

        clearInputVal = clearLeadingZeros(clearInputVal);
        const num = Number(clearInputVal);
        setValue(num);
        setRawValue(num === 0 ? clearInputVal : addPrefix(clearInputVal, _prefix));

        onChange(num);
    };

    return (
        <input
            className="MoneyInput"
            value={rawValue}
            onChange={onInput}
            autoFocus={autofocus}
            placeholder="0.00"
            type="tel"
        />
    );
};
