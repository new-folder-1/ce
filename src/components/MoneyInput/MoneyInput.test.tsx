import React from 'react';
import { shallow } from 'enzyme';

import { MoneyInput } from './MoneyInput';

describe('MoneyInput', () => {
    let onChange: jest.Mock;
    beforeEach(() => {
        onChange = jest.fn();
    });

    it('snapshot renders', () => {
        const component = shallow(
            <MoneyInput
                prefix="+"
                amount={200}
                onChange={onChange}
            />
        );

        expect(component).toMatchSnapshot();
    });

    //todo
    it('adds prefix when input changes', () => {
        const component = shallow(
            <MoneyInput
                prefix="+"
                onChange={onChange}
            />
        );

        const input = component
            .find('input');

        input.at(0).simulate('change', {
            currentTarget: { value: '2' }
        });

        expect(input).toMatchSnapshot()
    });
});
