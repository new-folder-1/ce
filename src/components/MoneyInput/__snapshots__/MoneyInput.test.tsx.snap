// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoneyInput adds prefix when input changes 1`] = `
<input
  className="MoneyInput"
  onChange={[Function]}
  placeholder="0.00"
  type="tel"
  value=""
/>
`;

exports[`MoneyInput snapshot renders 1`] = `
<input
  className="MoneyInput"
  onChange={[Function]}
  placeholder="0.00"
  type="tel"
  value="200"
/>
`;
